From 69b5618353d374a1356d92926dc129410ff74e30 Mon Sep 17 00:00:00 2001
From: Danny Lin <danny@kdrag0n.dev>
Date: Tue, 17 Aug 2021 21:24:33 -0700
Subject: [PATCH 26/34] gmscompat: Don't use Context to get device-protected
 data directory

Not all packages have data directories:

FATAL EXCEPTION: AsyncTask #2
Process: system:ui, PID: 8982
java.lang.RuntimeException: An error occurred while executing doInBackground()
	at android.os.AsyncTask$4.done(AsyncTask.java:415)
	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
	at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
	at java.util.concurrent.FutureTask.run(FutureTask.java:271)
	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:305)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:923)
Caused by: java.lang.RuntimeException: No data directory found for package android
	at android.app.ContextImpl.getDataDir(ContextImpl.java:2560)
	at com.android.internal.gmscompat.dynamite.client.DynamiteContext.<init>(DynamiteContext.java:58)
	at com.android.internal.gmscompat.dynamite.GmsDynamiteHooks.getClientContext(GmsDynamiteHooks.java:64)
	at com.android.internal.gmscompat.dynamite.GmsDynamiteHooks.loadAssetsFromPath(GmsDynamiteHooks.java:93)
	at android.content.res.ApkAssets.loadFromPath(ApkAssets.java:143)
	at android.app.ResourcesManager.loadApkAssets(ResourcesManager.java:374)
	at android.app.ResourcesManager.access$000(ResourcesManager.java:67)
	at android.app.ResourcesManager$ApkAssetsSupplier.load(ResourcesManager.java:146)
	at android.app.ResourcesManager.createApkAssetsSupplierNotLocked(ResourcesManager.java:833)
	at android.app.ResourcesManager.getResources(ResourcesManager.java:937)
	at android.app.ActivityThread.getTopLevelResources(ActivityThread.java:2226)
	at android.app.ApplicationPackageManager.getResourcesForApplication(ApplicationPackageManager.java:1682)
	at com.android.internal.app.ResolverListAdapter$TargetPresentationGetter.getIconBitmap(ResolverListAdapter.java:920)
	at com.android.internal.app.ResolverListAdapter$ActivityInfoPresentationGetter.getIconBitmap(ResolverListAdapter.java:843)
	at com.android.internal.app.ResolverListAdapter$TargetPresentationGetter.getIcon(ResolverListAdapter.java:908)
	at com.android.internal.app.ResolverListAdapter$ActivityInfoPresentationGetter.getIcon(ResolverListAdapter.java:843)
	at com.android.internal.app.ResolverListAdapter.loadIconForResolveInfo(ResolverListAdapter.java:599)
	at com.android.internal.app.ResolverListAdapter$LoadIconTask.doInBackground(ResolverListAdapter.java:782)
	at com.android.internal.app.ResolverListAdapter$LoadIconTask.doInBackground(ResolverListAdapter.java:769)
	at android.os.AsyncTask$3.call(AsyncTask.java:394)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	... 4 more

Use the underlying Environment methods instead to construct a
device-protected data path for GMs without relying on error-prone
Context methods.

Change-Id: I180cfe464f80a55ac534e150199ba1708841d13d
---
 .../gmscompat/dynamite/client/DynamiteContext.java    | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/core/java/com/android/internal/gmscompat/dynamite/client/DynamiteContext.java b/core/java/com/android/internal/gmscompat/dynamite/client/DynamiteContext.java
index 10d99679bd25..4fd3e06f996e 100644
--- a/core/java/com/android/internal/gmscompat/dynamite/client/DynamiteContext.java
+++ b/core/java/com/android/internal/gmscompat/dynamite/client/DynamiteContext.java
@@ -21,6 +21,7 @@ import android.content.Context;
 import android.content.Intent;
 import android.content.IntentFilter;
 import android.os.Bundle;
+import android.os.Environment;
 import android.os.Handler;
 import android.os.HandlerThread;
 import android.os.IBinder;
@@ -31,6 +32,8 @@ import com.android.internal.gmscompat.GmsInfo;
 import com.android.internal.gmscompat.dynamite.server.FileProxyProvider;
 import com.android.internal.gmscompat.dynamite.server.IFileProxyService;
 
+import java.io.File;
+
 /** @hide */
 public final class DynamiteContext {
     public static final String TAG = "GmsCompat/DynamiteClient";
@@ -53,10 +56,10 @@ public final class DynamiteContext {
     public DynamiteContext(Context context) {
         this.context = context;
 
-        // Use our own context and replace the package name to avoid ApkAssets recursion when
-        // lazy-creating a DynamiteContext in the ApkAssets hook
-        this.gmsDataPrefix = context.createDeviceProtectedStorageContext().getDataDir().getPath()
-                .replace(context.getPackageName(), GmsInfo.PACKAGE_GMS) + "/";
+        // Get data directory path without using package context or current data dir, since not all
+        // packages have data directories and package context causes recursion in ApkAssets
+        File userDe = Environment.getDataUserDeDirectory(null, context.getUserId());
+        gmsDataPrefix = userDe.getPath() + '/' + GmsInfo.PACKAGE_GMS + '/';
     }
 
     public ModuleLoadState getState() {
-- 
2.33.0

